src/managed/Microsoft.Extensions.DependencyModel/ArrayBufferWriter.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.DependencyModel
+{
+    internal class ArrayBufferWriter : IBufferWriter<byte>, IDisposable
+    {
+        private byte[] _rentedBuffer;
+        private int _written;
+
+        private const int MinimumBufferSize = 256;
+
+        public ArrayBufferWriter(int initialCapacity = MinimumBufferSize)
+        {
+            if (initialCapacity <= 0)
+                throw new ArgumentException(nameof(initialCapacity));
+
+            _rentedBuffer = ArrayPool<byte>.Shared.Rent(initialCapacity);
+            _written = 0;
+        }
+
+        public async Task CopyToAsync(Stream stream, CancellationToken cancellationToken = default)
+        {
+            if (_rentedBuffer == null)
+                throw new ObjectDisposedException(nameof(ArrayBufferWriter));
+
+            if (stream == null)
+                throw new ArgumentNullException(nameof(stream));
+
+            await stream.WriteAsync(_rentedBuffer, 0, _written, cancellationToken).ConfigureAwait(false);
+
+            _rentedBuffer.AsSpan(0, _written).Clear();
+            _written = 0;
+        }
+
+        internal void CopyTo(Stream stream)
```

Comment:
(nit) why have some methods `internal` and some `public` on an `internal` class?

Reply:
This particular API was made internal mainly because we likely won't have this synchronous API public on ArrayBufferWriter if we decide to ship it from corefx (only the async one would be public). All others should be public or private. I don't mind making it public if there is a need for it.